### ISSUE TITLE
[ASP.NET Core] System.InvalidOperationException on incoming web socket requests.

### ISSUE THREAD

**Issue by armingerten** [Microsoft/aspnet-api-versioning#342]
When using ASP.NET API versioning with web sockets, a System.InvalidOperationException is thrown when trying to add API version response headers.

Steps to reproduce: 

1. Configure API versionen
```csharp
public IServiceProvider ConfigureServices(IServiceCollection services)
{
    services..AddVersionedApiExplorer();
    services.AddApiVersioning(o =>
    {
        o.AssumeDefaultVersionWhenUnspecified = true;
        o.DefaultApiVersion = new ApiVersion(1, 0);
        o.ReportApiVersions = true;
    });
    return services;
}
```

2. Start application locally using the following controller / action:
```csharp
[ApiVersion("1.0")]
public class  MyController : Controller
{
    [HttpGet]
    public async Task OpenWebSocketConnectionAsync()
    {
        if (HttpContext.WebSockets.IsWebSocketRequest)
        {
            var webSocket = await HttpContext.WebSockets.AcceptWebSocketAsync();
            while (webSocket.IsOpen())
            {
                WebSocketReceiveResult result = await webSocket.ReceiveAsync(segment, CancellationToken.None);
                // Some business logic ...
            }
        }
    }
}
```
3. Try establishing a websocket request (and close it prematurely). 
```bash
curl --include --header "Connection:Upgrade" --header "Upgrade: websocket" http://localhost:8080
```

=> This will cause the following exception:
```
System.InvalidOperationException: Headers are read-only, response has already started.
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowHeadersReadOnlyException()
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.System.Collections.Generic.IDictionary<System.String,Microsoft.Extensions.Primitives.StringValues>.Add(String key, StringValues value)
at Microsoft.AspNetCore.Mvc.Versioning.DefaultApiVersionReporter.Report(IHeaderDictionary headers, ApiVersionModel apiVersionModel)
at Microsoft.AspNetCore.Mvc.Filters.ActionFilterAttribute.OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextExceptionFilterAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ExceptionContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

One possible solution would be to skip the reporting when dealing with `HttpContext.WebSockets.IsWebSocketRequest` or `HttpContext.Response.HasStarted` ?

I am using the following software versions.
- .NET core version: 2.1
- Microsoft.AspNetCore.Mvc.Versioning: 2.1

**Comment by commonsensesoftware** [comment]
This issue should be resolved in package version [2.3.0](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Versioning/2.3.0) as part of the fix for #273 and #290. My recommendation is to upgrade the package reference. If that doesn't fix your issue, I'm happy to investigate further. Thanks.

**Comment by armingerten** [comment]
Thanks for the quick reply! 

I just updated the references and re-tested:
- Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer: 2.2.0
- Microsoft.AspNetCore.Mvc.Versioning: 2.3.0

The issue seems to be fixed, so I am closing this - thank you!
